### PR TITLE
RavenDB-19703: Null keys support.

### PR DESCRIPTION
**Pull request by redknightlois**
https://issues.hibernatingrhinos.com/issue/RavenDB-19703/Allow-Corax-to-index-null-characters

For review purposes only. @ayende @maciejaszyk 

@arekpalinski can you run this indexing process on a copy of the live environments?

### CODE REVIEW

**Comment by maciejaszyk** [comment]
https://github.com/ravendb/ravendb/blob/2a146dd4d9698019a21004dc05dd91a1d3d1fc0e/src/Corax/IndexWriter.cs#L1404-L1407

You have to remove this guardian.

File: src/Sparrow.Server/Compression/Encoder3Gram.cs [hunk 1]
```diff
@@ -237,35 +188,75 @@ private string ToBinaryString(short sBuf, int sLen)
...
+        public void DecodeBatch<TSampleEnumerator, TOutputEnumerator>(in TSampleEnumerator data, Span<int> outputSize, in TOutputEnumerator outputBuffers)
             where TSampleEnumerator : struct, IReadOnlySpanIndexer
             where TOutputEnumerator : struct, ISpanIndexer
         {
-            fixed (Interval3Gram* table = EncodingTable)
+            ref Interval3Gram tableRef = ref EncodingTable[0];
+            var tree = BinaryTree<short>.Open(_state.DecodingTable);
+
+            byte* auxBuffer = stackalloc byte[4];
+            for (int i = 0; i < data.Length; i++)
```
Comment (arekpalinski) [inline src/Sparrow.Server/Compression/Encoder3Gram.cs:198]: @redknightlois please check `Unreachable code detected` warning here

**Comment by arekpalinski** [comment]
@redknightlois Sure, I can do it but please check the comments first

**Comment by arekpalinski** [comment]
@redknightlois @maciejaszyk can we have some tests using client API? For example to cover the case that we have in the issue description

File: src/Voron/Data/CompactTrees/CompactTree.cs [hunk 2]
```diff
@@ -2057,8 +2057,8 @@ private void SearchInCurrentPage(EncodedKey key, ref CursorState state)
                 if (match >= 0)
                 {
                     bot += mid;
-                    if (match == 0)
-                        goto IsMatch;
+                    if (match == 0 && keyLength == encodedKeyLength)
+                        goto IsExactMatch;
                 }
```
Comment (redknightlois) [inline src/Voron/Data/CompactTrees/CompactTree.cs]: @ayende @maciejaszyk I need a review here... Is it really a bug? Does it trigger only after doing the change to use encoded keys? How didn't we trigger this before?

Comment (ayende) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: Pretty sure that `match == 0` can happen _only_ if the length matches

Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: No, we are doing `Math.Min(keyLength, encodedKeyLength)` therefore they can still match but one been bigger than the other.  Which makes me wonder why we have coded it like this, that should be handled by `AdvMemory.CompareInline`. Am I right?

**Comment by arekpalinski** [comment]
Please note that test are failing

File: src/Voron/Data/CompactTrees/CompactTree.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: I'm not familiar with `CompactTree` implementation. Thus it is hard to tell if it's a bug or not.

Comment (ayende) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: Yes, this is needed, see here:
https://github.com/ravendb/ravendb/blob/599f19a74b39f3f2751f93df5c9f0a2373f5abf6/src/Voron/Data/CompactTrees/CompactTree.cs#L2061

We are checking the prefix for the start, then we compare the size.
However... if we have two values that are prefix of one another, and we may find the first one (that isn't a match) and then miss the other one that is a match

Comment (redknightlois) [reply src/Voron/Data/CompactTrees/CompactTree.cs]: Yes, that is exactly what I believe it was happening.

**Comment by redknightlois** [comment]
OK. It seems to work now. Waiting to run the slow tests.

**Pull request merged**